README.rst
================
django-chunkator
================


Chunk large QuerySets into small chunks, and iterate over them without killing
your RAM.

.. image:: https://travis-ci.org/peopledoc/django-chunkator.svg

Tested with all the combinations of:

* Python: 3.6, 3.7, 3.8
* Django: 2, 2.1, 2.2, 3.0, master

Usage
=====

.. code:: python

    from chunkator import chunkator
    for item in chunkator(LargeModel.objects.all(), 200):
        do_something(item)

This tool is intended to work on Django querysets.

Your model **must** define a ``pk`` field (this is done by default, but
sometimes it can be overridden) and this pk has to be unique. ``django-
chunkator`` has been tested with PostgreSQL and SQLite, using regular PKs and
UUIDs as primary keys.

You can also use ``values()``:

.. code:: python

    from chunkator import chunkator
    for item in chunkator(LargeModel.objects.values('pk', 'name'), 200):
        do_something(item)

.. important::

    If you're using ``values()`` you **have** to add at least your "pk" field
    to the values, otherwise, the chunkator will throw a
    ``MissingPkFieldException``.

.. warning::

    This will not **accelerate** your process. Instead of having one BIG query,
    you'll have several small queries. This will save your RAM instead, because
    you'll not load a huge queryset result before looping on it.


If you want to manipulate the pages directly, you can use `chunkator_page`:

.. code:: python

    from chunkator import chunkator_page
    queryset = LargeModel.objects.all().values('pk')
    for page in chunkator_page(queryset, 200):
        launch_some_task([item['pk'] for item in page])


FAQ
===

- How is django-chunkator different from Django's `iterator <https://docs.djangoproject.com/en/2.2/ref/models/querysets/#django.db.models.query.QuerySet.iterator>`_?

If you have server-side cursors (using Postgres or Oracle & not setting `DISABLE_SERVER_SIDE_CURSORS`), then the main difference is that the cursor is in the hands of the application instead of the server. It really depends on your constraints, but sometimes server side cursors can put too much strains on your DB.

If you don't have server-side cursors, then chunkator will allow you to iterate over your queryset by batch, without relying on LIMIT/OFFSET. The problem with LIMIT/OFFSET is that computing a large offset (when you're at the end of your queryset) requires the DB to go through all the previous entries. With large tables this can be a huge issue.

- Will django-chunkator preserve the ordering on my querysets?

No, it orders the queryset by pk. However you could do the same thing than chunkator with another field, given that it's unique and not nullable, see `here <https://github.com/peopledoc/django-chunkator/blob/master/chunkator/__init__.py#L27-L33>`_ for more details.


License
=======

MIT License.


CHANGELOG.md
# Changelog

## master (unreleased)

* Drop Python 3.5 support (#50).

## 2.0.0 (2020-06-22)

- Confirm support for Django 3.0 (#41).
- Remove `six` from the requirements (#43).
- Drop Python 2 Support (#37).
- Drop Django 1.11 support (#40).

## 1.5.0 (2020-02-05)

- Confirm support for Django 2.2 (#33).
- Add Python 3.8 support (#34).

## 1.4.0 (2018-12-17)

- Django 2 support
- Drop support for Django <1.11 and Python <3.5 (we keep 2.7 still)
- Confirm support for Django 2.1
- Add python 3.7 support

## 1.3.1 (2018-05-14)

- Changing Github Organization from ``novafloss`` to ``peopledoc`` (#24),
- Added documentation to install and bootstrap the demo app (#26),
- Removed `PendingDeprecationWarning` due to `StopIteration` usage (#23).

## 1.3.0 (2017-09-14)

- Run tox with Django 1.11 (#22).
- Run tests with the Django master version (#22).

## 1.2.0 (2017-07-03)

- Deal with OneToOne Primary Keys (#19).
- Expose the util function ``chunkator_page`` (#21).

## 1.1.0 (2017-03-25)

- Drop Django 1.6 and 1.7 support (#15).
- Add Django 1.10 support (#18).
- Add Python 3.5 support (#17).

## 1.0.1 (2016-08-26)

**Warning**: this version will be the last to support Django 1.6 or 1.7. Starting of the next version, we'll only support 1.8+.

- Added Django 1.9 support (#9).
- Python 3.4 jobs are being tested in travis (#10).
- Deprecate Django 1.6 and 1.7 support (#9).

## 1.0.0 (2016-06-23)

* Django 1.8 compatibility, thank you ``tox`` (#5).
* Compatible with Python 3 (tested with Python 3.4), thx @wo0dyn (#6).
* Various fixes and added a proper MIT License file (#8).

## v0.0.5

(released on 2015-06-18)

* Bugfix: removed multiple "pk > 'x'" condition statements (#4).

## v0.0.4

(released on 2015-03-30)

* Allow the usage of ``values()`` instead of the usual queryset (#2 and #3)

## v0.0.3

(released on 2015-03-19)

* allow "non-integer" pk (#1), such as UUIDs.


## v0.0.2

(released on 2015-03-19)

* tox and travis setup,
* a few more tests

## v0.0.1

(released on 2015-03-18)

* Proof of concept release, we can iterate over large querysets chunk by chunk,
* include demo test suite,


demo/demo_chunkator/tests.py
from io import StringIO

from django.test import TestCase

from chunkator import MissingPkFieldException
from chunkator import chunkator
from chunkator import chunkator_page
from demo_chunkator.models import Book
from demo_chunkator.models import Cover
from demo_chunkator.models import Profile
from demo_chunkator.models import User


class ChunkatorTestCase(TestCase):

    def setUp(self):
        super(ChunkatorTestCase, self).setUp()
        for nb in range(20):
            Book.objects.create(
                title="Title #{}".format(nb),
                author="Author #{}".format(nb)
            )

    def test_chunks_queryset(self):
        # small step
        chunks = chunkator(Book.objects.all(), 1)
        result = []
        for item in chunks:
            self.assertTrue(isinstance(item, Book))
            result.append(item.pk)
        self.assertEqual(len(result), 20)
        self.assertEqual(len(result), len(set(result)))  # no duplicates

        result = []
        # larger chunks
        chunks = chunkator(Book.objects.all(), 10)
        for item in chunks:
            self.assertTrue(isinstance(item, Book))
            result.append(item.pk)
        self.assertEqual(len(result), 20)
        self.assertEqual(len(result), len(set(result)))  # no duplicates

        result = []
        # larger than QS chunks
        chunks = chunkator(Book.objects.all(), 50)
        for item in chunks:
            self.assertTrue(isinstance(item, Book), "{}".format(item))
            result.append(item.pk)
        self.assertEqual(len(result), 20)
        self.assertEqual(len(result), len(set(result)))  # no duplicates

    def test_chunks_numqueries(self):
        # Make sure we only run 2 queries
        # One for each slice
        with self.assertNumQueries(2):
            chunks = chunkator(Book.objects.all(), 12)
            for item in chunks:
                self.assertTrue(isinstance(item, Book))

        # Make sure we only run 3 queries
        # One for each slice, plus the "empty" one
        with self.assertNumQueries(3):
            chunks = chunkator(Book.objects.all(), 10)
            for item in chunks:
                self.assertTrue(isinstance(item, Book))

        # If the chunk size is bigger than the number of items, only one query.
        with self.assertNumQueries(1):
            chunks = chunkator(Book.objects.all(), 1000)
            for item in chunks:
                self.assertTrue(isinstance(item, Book))


class ChunkatorUUIDTestCase(TestCase):

    def setUp(self):
        super(ChunkatorUUIDTestCase, self).setUp()
        User.objects.create(name='Terry Pratchett')
        User.objects.create(name='Iain Banks')

    def test_chunk_uuid(self):
        result = []
        chunks = chunkator(User.objects.all(), 10)
        for item in chunks:
            self.assertTrue(isinstance(item, User))
            result.append(item.pk)
        self.assertEqual(len(result), 2)
        self.assertEqual(len(result), len(set(result)))  # no duplicates


class ChunkatorOrderTestCase(TestCase):
    def setUp(self):
        super(ChunkatorOrderTestCase, self).setUp()
        book = Book.objects.create(
            title="The Player of Games",
            author="Banks, Iain"
        )
        Cover.objects.create(book=book, code='player')
        book = Book.objects.create(
            title="Guards! Guards!",
            author="Pratchett, Terry"
        )
        Cover.objects.create(book=book, code='guards')

        user = User.objects.create(name='Wonder Woman')
        Profile.objects.create(user=user, avatar='wonderful')
        user = User.objects.create(name='Wolverine')
        Profile.objects.create(user=user, avatar='wolfy')

    def test_order_by_default(self):
        items = list(chunkator(Book.objects.all(), 10))
        self.assertEqual(items[0].pk, 1)
        self.assertEqual(items[1].pk, 2)

        uuids = sorted(User.objects.values_list('pk', flat=True))
        items = list(chunkator(User.objects.all(), 10))
        self.assertEqual(items[0].pk, uuids[0])
        self.assertEqual(items[1].pk, uuids[1])

    def test_order_by_with_onetoone_pk(self):
        items = list(chunkator(Cover.objects.all(), 10))
        self.assertEqual(items[0].pk, 1)
        self.assertEqual(items[1].pk, 2)

        items = list(chunkator(Cover.objects.all(), 10))
        self.assertEqual(items[0].pk, 1)
        self.assertEqual(items[1].pk, 2)


class ChunkatorValuesTestCase(TestCase):

    def setUp(self):
        super(ChunkatorValuesTestCase, self).setUp()
        User.objects.create(name='Wonder Woman')
        User.objects.create(name='Wolverine')

    def test_chunk_uuid(self):
        result = []
        chunks = chunkator(User.objects.all().values("pk", "name"), 10)
        for item in chunks:
            self.assertTrue(isinstance(item, dict))
            result.append(item['pk'])
        self.assertEqual(len(result), 2)
        self.assertEqual(len(result), len(set(result)))  # no duplicates

    def test_chunk_missing_pk(self):
        with self.assertRaises(MissingPkFieldException):
            result = chunkator(User.objects.all().values("name"), 10)
            next(result)


class ChunkatorWhereTest(TestCase):
    def setUp(self):
        super(ChunkatorWhereTest, self).setUp()
        User.objects.create(name='Wonder Woman')
        User.objects.create(name='Wolverine')
        User.objects.create(name='ChuckNorris')

    def test_query_log(self):
        query_log_output = StringIO()
        qs = User.objects.all()
        # We loop here only to dig into the generator and force execution
        for item in chunkator(qs, 1, query_log=query_log_output):
            _ = item  # noqa
        contents = query_log_output.getvalue()
        query_log_output.close()
        queries = contents.split('\n')
        self.assertEqual(len(queries), 5, queries)
        queries = queries[:4]  # the last one is empty string
        for query in queries:
            # Should be 0 for the first query
            # Should occur once for other queries
            self.assertTrue(query.count('."uuid" >') <= 1, query)


class ChunkatorPageTest(TestCase):
    def setUp(self):
        super(ChunkatorPageTest, self).setUp()
        for nb in range(3):
            Book.objects.create(
                title="Title #{}".format(nb),
                author="Author #{}".format(nb)
            )

    def test_chunkator_page(self):
        qs = Book.objects.all().values('pk')
        pages = list(chunkator_page(qs, 2))
        self.assertEqual(pages, [[{'pk': 1}, {'pk': 2}], [{'pk': 3}]])


chunkator/__init__.py
"""
Toolbox for chunking / slicing querysets
"""


class MissingPkFieldException(Exception):
    """
    Exception raised when the "pk" field is missing in the query.

    (when using ``values()``).
    """
    pass


def chunkator_page(source_qs, chunk_size, query_log=None):
    """
    Yield pages of a queryset.
    """
    pk = None
    has_fields = hasattr(source_qs, '_fields') and source_qs._fields
    if has_fields:
        if "pk" not in source_qs._fields:
            raise MissingPkFieldException("The values() call must include the `pk` field")  # noqa

    field = source_qs.model._meta.pk
    # set the correct field name:
    # for ForeignKeys, we want to use `model_id` field, and not `model`,
    # to bypass default ordering on related model
    order_by_field = field.attname

    source_qs = source_qs.order_by(order_by_field)
    queryset = source_qs
    while True:
        if pk:
            queryset = source_qs.filter(pk__gt=pk)
        page = queryset[:chunk_size]
        if query_log is not None:
            query_log.write('{page.query}\n'.format(page=page))
        page = list(page)
        nb_items = len(page)

        if nb_items == 0:
            return

        last_item = page[-1]
        # source_qs._fields exists *and* is not none when using "values()"
        if has_fields:
            pk = last_item["pk"]
        else:
            pk = last_item.pk

        yield page

        if nb_items < chunk_size:
            return


def chunkator(source_qs, chunk_size, query_log=None):
    """
    Yield over a queryset by chunks.

    This method does not involve counting elements or measuring the iterable
    length. We're saving at least a ``count()`` query on QuerySets, or a
    CPU-and-RAM-consuming ``len(queryset)`` query.
    """
    for page in chunkator_page(source_qs,
                               chunk_size,
                               query_log=query_log):
        for item in page:
            yield item


demo/demo_chunkator/migrations/0001_initial.py
# -*- coding: utf-8 -*-
# Generated by Django 1.11 on 2018-05-14 04:22
from __future__ import unicode_literals

from django.db import migrations, models
import django.db.models.deletion
import uuid


class Migration(migrations.Migration):

    initial = True

    dependencies = [
    ]

    operations = [
        migrations.CreateModel(
            name='Book',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('title', models.CharField(max_length=200)),
                ('author', models.CharField(max_length=200)),
            ],
            options={
                'ordering': ['title'],
            },
        ),
        migrations.CreateModel(
            name='User',
            fields=[
                ('uuid', models.UUIDField(default=uuid.uuid4, primary_key=True, serialize=False)),
                ('name', models.CharField(max_length=100)),
            ],
            options={
                'ordering': ['name'],
            },
        ),
        migrations.CreateModel(
            name='Cover',
            fields=[
                ('book', models.OneToOneField(on_delete=django.db.models.deletion.CASCADE, primary_key=True, serialize=False, to='demo_chunkator.Book')),
                ('code', models.CharField(max_length=20)),
            ],
        ),
        migrations.CreateModel(
            name='Profile',
            fields=[
                ('user', models.OneToOneField(on_delete=django.db.models.deletion.CASCADE, primary_key=True, serialize=False, to='demo_chunkator.User')),
                ('avatar', models.CharField(max_length=100)),
            ],
        ),
    ]
